[MXNET-80] Fix average pooling kernel size assignment error

CoinCheung opened the pull request:
## Description ##
The pooling operator still complains about not assigning the kernel size when the "global_pool" parameter is assigned to be "True";

## Checklist ##
### Essentials ###
- [x] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

sxjscience commented:
Thanks for submitting this fix! You can remove the `kernel=` phrases in the test cases (https://github.com/apache/incubator-mxnet/blob/master/tests/python/gpu/test_operator_gpu.py#L914-L935). Also, you may run `make pylint` to check the lint locally.

sxjscience commented on src/operator/nn/pooling-inl.h:
```diff
@@ -178,6 +178,11 @@ class PoolingOp {
+  // check if filter size assigned correctly
+  if (param.global_pool == false) {
+    CHECK_GT(param.kernel.ndim(), 0U)
+        << "A positive number must be assigned as filter size";
```
Need a better error message.

CoinCheung replied:
How about "Must assign a positive kernel size"?

sxjscience replied:
I think we can use "You need to set the kernel size if global pooling is not used".

sxjscience commented:
The error code show that the change somehow breaks the cpp package. I need to investigate the status of the CPP package.

sxjscience commented:
@CoinCheung The error is due to that `pool_type` has no default value. We should change this line to default it max https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/pooling-inl.h#L63.

TaoLv commented on tests/python/gpu/test_operator_gpu.py:
```diff
@@ -904,86 +904,87 @@ def test_1d_pooling(pool_type):
+
         ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
         sym_list.append(mx.sym.Pooling_v1(kernel=kernel, pad=pad, stride=stride, pool_type=pool_type,
                                        pooling_convention=pooling_convention, global_pool=True, name='pool'))
```
Please help to fix the indent.

TaoLv commented on tests/python/gpu/test_operator_gpu.py:
```diff
@@ -904,86 +904,87 @@ def test_1d_pooling(pool_type):
+
         ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
         sym_list.append(mx.sym.Pooling_v1(kernel=kernel, pool_type=pool_type,
                                        pooling_convention=pooling_convention, global_pool=True, name='pool'))
```
Please help to fix the indent.

CoinCheung replied:
Do I need to remove the blank lines? I only removed the kernel parameter assignment did not touch these white lines.

sxjscience replied:
You can do it if you have time. It should be due to the difference between windows and unix.

TaoLv commented on tests/python/gpu/test_operator_gpu.py:
```diff
@@ -904,86 +904,87 @@ def test_1d_pooling(pool_type):
-    
+
         ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
         sym_list.append(mx.sym.Pooling_v1(kernel=kernel, pad=pad, stride=stride, pool_type=pool_type,
```
Do you think we should also remove the `kernel` phrases in 2d tests? I wrote this test case to make sure that there is a same behavior for global pooling w/ and w/o `pad` and `stride`. It seems `kernel` should also be checked in the same way.

sxjscience replied:
But `kernel` is not used if global_pool=True.

TaoLv replied:
We cannot prevent user from inputing a `kernel` to global pooling. Kernel size will be reset to image shape if `global_pool` is true. Code here: https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/pooling-inl.h#L143

sxjscience replied:
I see. We should then use add back one test case that uses the kernel argument.

CoinCheung replied:
So shall I remove kernel only in the "even number" test cases and leave the odd test case with their kernel? Such as:
```
        ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
        sym_list.append(mx.sym.Pooling(kernel=kernel, pad=pad, stride=stride, pool_type=pool_type,  # keep the kernel for checking
                                       pooling_convention=pooling_convention, global_pool=True, name='pool'))

        ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
        sym_list.append(mx.sym.Pooling(pool_type=pool_type,  # remove kernel along with the missing pad and stride
                                       pooling_convention=pooling_convention, global_pool=True, name='pool'))

```

sxjscience replied:
Sounds good.

TaoLv replied:
Maybe we shouldn't remove the current test points. We can add new test points into the test case. For me, the 3 test points below should have same result.
```
ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
sym_list.append(mx.sym.Pooling(kernel=kernel, pad=pad, stride=stride, pool_type=pool_type,
                               pooling_convention=pooling_convention, global_pool=True, name='pool'))

ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
sym_list.append(mx.sym.Pooling(kernel=kernel, pool_type=pool_type,
                               pooling_convention=pooling_convention, global_pool=True, name='pool'))
# below is new test point
ctx_list.append({'ctx': mx.cpu(0), 'pool_data': data, 'type_dict': {'pool_data': np.float32}})
sym_list.append(mx.sym.Pooling(pool_type=pool_type,
                               pooling_convention=pooling_convention, global_pool=True, name='pool'))
```

sxjscience replied:
@CoinCheung, we can then change to append new tests in the list instead of replacing the current tests. We could change the code following the comment by @TaoLv

CodingCat commented:
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

sxjscience commented on src/operator/nn/pooling-inl.h:
```diff
@@ -56,11 +56,11 @@ struct PoolingParam : public dmlc::Parameter<PoolingParam> {
+    DMLC_DECLARE_FIELD(pool_type).set_default(pool_enum::kMaxPooling)  // add default pooling method
     .add_enum("max", pool_enum::kMaxPooling)
     .add_enum("avg", pool_enum::kAvgPooling)
     .add_enum("sum", pool_enum::kSumPooling)
```
I realized that we need to change the order of the DMLC_DECLARE_FIELD here. In the original version, the parameters that do not have default values will be set first and then goes the params with default values. So the order will be `kernel`, `pool_type`, `global_pool`, `cudnn_off`, ... After we add default values to `kernel` and `pool_type`, the order becomes `global_pool`, `cudnn_off`, `kernel`, `pool_type`, ... Thus, the way to solve the problem is to change the order:
```c++
    DMLC_DECLARE_FIELD(kernel).set_default(TShape())  // add default value here
    .enforce_nonzero()
    .describe("Pooling kernel size: (y, x) or (d, y, x)");

    DMLC_DECLARE_FIELD(pool_type).set_default(pool_enum::kMaxPooling)  // add default pooling method
    .add_enum("max", pool_enum::kMaxPooling)
    .add_enum("avg", pool_enum::kAvgPooling)
    .add_enum("sum", pool_enum::kSumPooling)
    .describe("Pooling type to be applied.");

    DMLC_DECLARE_FIELD(global_pool).set_default(false)
    .describe("Ignore kernel size, do global pooling based on current input feature map. ");

    DMLC_DECLARE_FIELD(cudnn_off).set_default(false)
    .describe("Turn off cudnn pooling and use MXNet pooling operator. ");
```

CoinCheung replied:
But I see in the original version that the order is: global_pool(false), cudnn_off(false), kernel(no default value), pool_type(no default value), ....

In the original version, "kernel" and "pool_type" do not have default value, but they go after "global_pool" and "cudnn_off" which are with default values.

https://github.com/apache/incubator-mxnet/blob/94f68fc8fd21611b7f5c148cb0e5d134efe58f87/src/operator/nn/pooling-inl.h#L59

CoinCheung replied:
Sorry, maybe I have misunderstood what you said, I will have a try.

CoinCheung replied:
I tried a few times, and it failed at this position:
https://github.com/apache/incubator-mxnet/blob/94f68fc8fd21611b7f5c148cb0e5d134efe58f87/src/operator/nn/pooling.cc#L55
But I do not understand why it requires stride and kernel have the same length.

sxjscience replied:
There's no need to check this if global_pool is turned on.

sxjscience commented:
@CoinCheung I've created the JIRA issue for you. See https://issues.apache.org/jira/browse/MXNET-80

sxjscience commented on src/operator/nn/pooling.cc:
```diff
@@ -54,11 +54,13 @@ static void PoolingParamParser(nnvm::NodeAttrs *attrs) {
+        << "stride and kernel should have the same length";
+    CHECK_EQ(param.pad.ndim(), param.kernel.ndim())
+        << "pad and kernel should have the same length";
+    attrs->parsed = std::move(param);
```
We still need to parse the param

sxjscience commented:
We may also need to revise the shape assignment logic: https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/pooling.cc#L103-L215

CoinCheung commented:
@sxjscience 
But I did not see logic error in this function. From the printed error message, I see no CHECK() failure triggered, and in every scenario with global_pool, the output shape is set to [-1,1] or [-1,1,1,] or [-1,1,1,1].

sxjscience commented:
I think it has not handled the case when kernel.ndim()=0

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: CoinCheung <notifications@github.com>
Sent: Tuesday, March 13, 2018 12:33:34 AM
To: apache/incubator-mxnet
Cc: Xingjian SHI; Mention
Subject: Re: [apache/incubator-mxnet] [MXNET-80] Fix average pooling kernel size assignment error (#10000)


@sxjscience<https://github.com/sxjscience>
But I did not see logic error in this function. From the printed error message, I see no CHECK() failure triggered, and in every scenario with global_pool, the output shape is set to [-1,1] or [-1,1,1,] or [-1,1,1,1].

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-mxnet/pull/10000#issuecomment-372572675>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE8D7lrKXOv4rxBdrwh-iqpHWKqseDbFks5td3ZNgaJpZM4Sd6lm>.

CoinCheung commented:
I tried but failed. So what should be the correct behavior when kernel.ndim() is 0? @sxjscience

sxjscience commented on src/operator/nn/pooling.cc:
```diff
@@ -98,9 +103,24 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
+          oshape[2] = 1;
+          oshape[3] = 1;
+          oshape[4] = 1;
+      }
```
Need to push the oshape to `out_shape`. See https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/pooling.cc#L163-L168.

sxjscience replied:
Also, you can use a for-loop instead of if.

CoinCheung replied:
There is another mismatch at this line:
https://github.com/apache/incubator-mxnet/blob/c9ec3118688c233a66ad847003a9e8d2d09e5952/src/operator/nn/pool.h#L690
and I have no idea how I could fix this as this function pool() have no global_pool as input parameter.

CoinCheung replied:
I feel it is sort of dangerous to modify function definitions and add a input parameter. Do we have other choices ?

sxjscience replied:
There is no need to revise this function. All you need to do is to prepare the correct initial values. You can check the logic here https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/pooling-inl.h#L135-L149 . If global_pool is used, kernel is set to be `TShape(ishape.data()+ishape.ndim()-param_.kernel.ndim(), ishape.data()+ishape.ndim())`, pad is set to be all 0 and stride is set to be all 1. You can write a function to do this. For example:
```c++
TShape kernel = param_.kernel;
TShape padding = param_.pad;
TShape stride = param_.stride;
if(param_.global_pool) {
   kernel = TShape(ishape.data()+ishape.ndim()-param_.kernel.ndim(), ishape.data()+ishape.ndim());
   padding = TShape(ishape.ndim() - 2);
   for(int i = 0; i < ishape.ndim() - 2; i++) {
      padding[i] = 0;
   }
   stride = TShape(ishape.ndim() - 2);
}
pool(s, in_data.dptr<DType>(), in_data.shape_, out_data.shape_,
     kernel,
     padding,
     stride,
     param_.pool_type, req, out_data.dptr<DType>());
```

sxjscience commented on src/operator/nn/pooling.cc:
```diff
@@ -98,9 +103,24 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
+  } else if (param.kernel.ndim() == 1) {
     CHECK_EQ(dshape.ndim(), 3U)
         << "Pooling: Input data should be 3D in (batch, channel, x)";
     if (param.global_pool) {
```
These `if` can be removed because the global_pool case is handled before.

sxjscience commented on src/operator/nn/pooling.cc:
```diff
@@ -46,15 +46,20 @@ static void PoolingParamParser(nnvm::NodeAttrs *attrs) {
+        << "stride and kernel should have the same length";
+    CHECK_EQ(param.pad.ndim(), param.kernel.ndim())
+        << "pad and kernel should have the same length";
+  }
```
Simplify the logic here. If you have used global_pool, there is no need to check the stride, kernel or pad.
```c++
if(! param.global_pool) {
// CHECK kernel, pad, stride
}
```

sxjscience commented on src/operator/nn/pooling.cc:
```diff
@@ -130,35 +153,30 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
+      oshape[3] = 1 +
+                  (dshape[3] + 2 * param.pad[1] - param.kernel[1]) /
+                      param.stride[1];
     } else {
```
The `if (param.global_pool)` is removed and the `else` should also be removed.

sxjscience replied:
Ignore this comment. I've misinterpreted the code.

CoinCheung commented on src/operator/nn/pooling-inl.h:
```diff
@@ -132,19 +132,21 @@ class PoolingOp {
+    TShape stride = param_.stride;
     if (param_.global_pool) {
-      for (index_t i = 0; i < padding.ndim(); i++) {
+      kernel = TShape(ishape.data()+ishape.ndim()-param_.kernel.ndim(), ishape.data()+ishape.ndim());
```
@sxjscience 
It still does not work even though kernel is set accordingly. By the way, I do not think the original implementation does wrong here, as it also considers the global_pool and resets the kernel.  
Also I am not familiar with the source codes and the codes are in such a mass that I failed to find the definition of ```TShape```. Would you please show me where ```TShape``` is defined?

piiswrong commented:
Changing the order of declare field is not allowed. It will break API

sxjscience commented:
@piiswrong I find that the params without default values are parsed first and then goes the params with default values.

sxjscience commented:
@piiswrong We changed the order because the CPP package could not compile without changing it.

sxjscience commented:
@CoinCheung Could you rebase to the current master? You can follow the guide in https://mxnet.incubator.apache.org/community/contribute.html?highlight=rebase and https://makandracards.com/makandra/527-squash-several-git-commits-into-a-single-commit

CoinCheung commented:
I tried but am not sure if this is right, please correct me if that is not you want.

sxjscience commented:
@CoinCheung We may stash the commits into 1 commit.

sxjscience commented on src/operator/nn/pool.h:
```diff
@@ -687,7 +687,7 @@ inline void pool(mshadow::Stream<cpu>* s, const DType* in_data, const TShape& is
     }
   } else {
-    LOG(FATAL) << "Unsupported " << kernel.ndim() << "-D pooling";
+    LOG(FATAL) << "Unsupported " << kernel.ndim() << "-D non-avg pooling";
```
Why non-avg here?

CoinCheung replied:
I used to misunderstand the logic there and I change it back now. It might be fine now.

sxjscience commented on src/operator/nn/pooling.cc:
```diff
@@ -98,28 +97,45 @@ static bool PoolingShape(const nnvm::NodeAttrs &attrs,
   if (dshape.ndim() == 0) return false;
-  if (param.kernel.ndim() == 1) {
+  if (param.global_pool) {
+      if (dshape.ndim() == 3) {
```
We can replace the if-clauses with a for-loop.

sxjscience commented on src/operator/pooling_v1-inl.h:
```diff
@@ -215,18 +217,21 @@ class PoolingV1Prop : public OperatorProperty {
-    if (param_.kernel.ndim() == 2) {
-      if (param_.stride.ndim() == 0) param_.stride = Shape2(1, 1);
-      if (param_.pad.ndim() == 0) param_.pad = Shape2(0, 0);
+    if (param_.global_pool) {
```
if (! param_.global_pool) {...

piiswrong commented:
You can't change the order. You can only add default values for all later args

CoinCheung commented:
@piiswrong I changed the order back, but it cannot pass the compiling, even though every arg has a default assignment. But it seems to work well in compiling and tests process if I switch the orders.

sxjscience commented:
@CoinCheung Would you test to see if `Pooling` can still be run with the original argument order? (In the python side). This means to not use `mx.nd.Pooling(x, kernel=(h, w), ...)`, but to directly use `mx.nd.Pooling(x, (h, w)...`

CoinCheung commented:
@sxjscience Ok, I will have a try. By the way, the latest commit failed the test at some place I do not quite understand. Would you tell me what happened there?

sxjscience commented:
Some flaky tests fail randomly.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: CoinCheung <notifications@github.com>
Sent: Wednesday, March 21, 2018 11:05:37 PM
To: apache/incubator-mxnet
Cc: Xingjian SHI; Mention
Subject: Re: [apache/incubator-mxnet] [MXNET-80] Fix average pooling kernel size assignment error (#10000)


@sxjscience<https://github.com/sxjscience> Ok, I will have a try. By the way, the latest commit failed the test at some place I do not quite understand. Would you tell me what happened there?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-mxnet/pull/10000#issuecomment-375190099>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE8D7m7y3yG-rDJJj0NVDSH_mEIbFSS8ks5tgz8xgaJpZM4Sd6lm>.

sxjscience commented:
@CoinCheung There's some problems with the CI. You need to rebase and stash the commits and use `git push --force`

CoinCheung commented:
@sxjscience I did nothing since the last commit. But the newly submitted code cannot be compiled.

sxjscience commented:
@CoinCheung It's not your fault. This is due to problems in the CI. You need to rebase against the master to trigger the build again. Here is a [good guide](https://makandracards.com/makandra/527-squash-several-git-commits-into-a-single-commit)

You can use the following commands:
```
git remote add apache https://github.com/apache/incubator-mxnet.git
git fetch apache
git rebase -i apache/master
```
You will see a file open up and change all the `push` into `s` and save&close the file.
After the rebase is finished, use `git push --force` to force update.

CoinCheung commented:
@sxjscience changing the order back will even not pass all the compiling. Shall I remove the /example and the /cpp-package directory to avoid these cpp errors?

sxjscience commented:
Have you tested whether the original python codes could still be run when we change the order?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: CoinCheung <notifications@github.com>
Sent: Thursday, March 22, 2018 7:14:08 PM
To: apache/incubator-mxnet
Cc: Xingjian SHI; Mention
Subject: Re: [apache/incubator-mxnet] [MXNET-80] Fix average pooling kernel size assignment error (#10000)


@sxjscience<https://github.com/sxjscience> changing the order back will even not pass all the compiling. Shall I remove the /example and the /cpp-package directory to avoid these cpp errors?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-mxnet/pull/10000#issuecomment-375517871>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE8D7lBZ75m7LRW5jMairjWVX4ocWsUVks5thFpwgaJpZM4Sd6lm>.

CoinCheung commented:
@sxjscience I tried and find it does not work. Given a kernel without indicating with ```kernel=``` will let python believe it is the first unassigned argument ```data```. It might work only if all the other arguments before ```kernel``` are given, because in this way, the kernel input without ```kernel=``` is assigned to the first unassigned argument.

sxjscience commented:
@CoinCheung Let's first change the order to make sure the test passed and I'll discuss with @piiswrong about whether it is acceptable.

CoinCheung commented:
@sxjscience I changed the order and it passed all the test.

sxjscience commented:
We need to either break the backward compatibility of the CPP package or the backward compatibility of the Python package. For CPP package, the Pooling layer is used like this https://github.com/apache/incubator-mxnet/blob/master/cpp-package/example/googlenet.cpp#L88-L89

CoinCheung commented:
We merely changed the order in the function body rather than the function declarations, and it did pass the  test. Why do we still need to break some compatibility?

sxjscience commented:
In the python side, we can call functions in two ways. 1)Using kwargs, i.e mx.nd.Pooling(data, kernel=kernel, pad=pad, ...) 2) Using args, i.e, mx.nd.Pooling(data, kernel, pad) Our current change will break the backward compatibility of the second case. However, I think the second case is seldom used and it’s okay to break the compatibility.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: CoinCheung <notifications@github.com>
Sent: Friday, March 23, 2018 7:42:19 PM
To: apache/incubator-mxnet
Cc: Xingjian SHI; Mention
Subject: Re: [apache/incubator-mxnet] [MXNET-80] Fix average pooling kernel size assignment error (#10000)


We merely changed the order in the function body rather than the function declarations, and it did pass the test. Why do we still need to break some compatibility?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-mxnet/pull/10000#issuecomment-375840984>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE8D7vIhmJyIk3rSb9JL3KPC2Sjv3gitks5thbKKgaJpZM4Sd6lm>.

CoinCheung commented:
Have we changed the order of python api args by changing the order in the C++ source code function body. Do you mean after the changing, the python api is also changed?
like changing python api from:
```
mxnet.symbol.Pooling(data=None, global_pool=_Null, cudnn_off=_Null, kernel=_Null, pool_type=_Null, pooling_convention=_Null, stride=_Null, pad=_Null, name=None, attr=None, out=None, **kwargs)
```
To
```
mxnet.symbol.Pooling(data=None,  kernel=_Null, pool_type=_Null, global_pool=_Null, cudnn_off=_Null, pooling_convention=_Null, stride=_Null, pad=_Null, name=None, attr=None, out=None, **kwargs)
```

sxjscience commented:
Yes.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: CoinCheung <notifications@github.com>
Sent: Sunday, March 25, 2018 6:08:29 PM
To: apache/incubator-mxnet
Cc: Xingjian SHI; Mention
Subject: Re: [apache/incubator-mxnet] [MXNET-80] Fix average pooling kernel size assignment error (#10000)


Have we changed the order of python api args by changing the order in the C++ source code function body. Do you mean after the changing, the python api is also changed?
like changing python api from:

mxnet.symbol.Pooling(data=None, global_pool=_Null, cudnn_off=_Null, kernel=_Null, pool_type=_Null, pooling_convention=_Null, stride=_Null, pad=_Null, name=None, attr=None, out=None, **kwargs)


To

mxnet.symbol.Pooling(data=None,  kernel=_Null, pool_type=_Null, global_pool=_Null, cudnn_off=_Null, pooling_convention=_Null, stride=_Null, pad=_Null, name=None, attr=None, out=None, **kwargs)


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/apache/incubator-mxnet/pull/10000#issuecomment-376019631>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AE8D7pqyHUoAu5FTipm3EOVgQTaX-Ifsks5tiD-MgaJpZM4Sd6lm>.

sxjscience commented:
@CoinCheung I just discussed it with Eric. We agree that the current way is acceptable. Please resolve the conflict and we can merge.

anirudh2290 commented:
@CoinCheung can you please resolve the conflicts.

CoinCheung commented:
It failed somewhere I don`t think associated with this pull request.

anirudh2290 commented:
@sxjscience @piiswrong Good to merge now ?

sxjscience commented:
Yes, good to merge. We need to mention the API change in the release note.
